docs/src/content/en/docs/agents/input-processors.mdx

```diff
@@ -0,0 +1,438 @@
+
+## Overview
+
+Input processors run sequentially before the agent sends messages to the language model. Each processor receives a context object and must return an array of `MastraMessageV2` objects. The returned array is then passed as input to the next processor in the chain.
+
+The context object contains:
+- **`messages`**: An array of `MastraMessageV2` objects representing the current messages in the pipeline. You can mutate this array directly and return it.
+- **`abort(reason?)`**: A function to terminate processing early with an optional reason.
+
+This design allows you to build sophisticated and predictable message processing pipelines with validation, transformation, and conditional logic.
+
+## Basic Usage
+
+### Creating an Input Processor
+
+To create an input processor, define an object that conforms to the `InputProcessor` interface.
+
+```typescript
+import { Agent, type InputProcessor, type MastraMessageV2 } from "@mastra/core/agent";
+
+const contentFilter: InputProcessor = {
+  name: 'content-filter',
+  process: async ({ messages, abort }) => {
+    // Check for inappropriate content in the messages array
+    const hasInappropriate = messages.some(msg => {
+      // A helper to extract text content is useful
+      const textContent = msg.content.parts
+        .filter(part => part.type === 'text')
+        .map(part => (part as { text: string }).text)
+        .join(' ');
+      
+      return textContent.includes('inappropriate');
+    });
+    
+    if (hasInappropriate) {
+      abort('Content validation failed');
+    }
+    
+    return messages;
+  },
+};
+```
+
+### Adding Processors to an Agent
+
+Add processors to your agent's configuration. They will be executed in the order they appear in the array.
+
+```typescript
+const agent = new Agent({
+  name: 'content-aware-agent',
+  instructions: 'You are a helpful assistant',
+  model: myLanguageModel,
+  inputProcessors: [contentFilter]
+});
+```
+
+## Message Manipulation
+
+You can directly mutate the `messages` array. After performing modifications, return the same array.
+
+### Adding Messages
+
+```typescript
+const messageEnhancer: InputProcessor = {
+  name: 'enhancer',
+  process: async ({ messages }) => {
+    const newMessage: MastraMessageV2 = {
+      id: crypto.randomUUID(),
+      role: 'assistant',
+      content: {
+        format: 2,
+        parts: [{ type: 'text', text: 'Please be concise in your response.' }]
+      },
+      createdAt: new Date()
+    };
+    
+    messages.push(newMessage);
+    
+    return messages;
+  },
+};
+```
+
+### Modifying Existing Messages
+
+Use a `for...of` loop to iterate through messages and update their content in place.
+
+```typescript
+const messageModifier: InputProcessor = {
+  name: 'modifier',
+  process: async ({ messages }) => {
```

Comment:
even though they can't modify system messages we should probably still pass them in as an arg - token limiter for example in memory needed to measure the system prompt even though it couldn't modify it